File: test/escapement/tui/transcript_test.clj
(ns escapement.tui.transcript-test
  "Unit specs for the SENT/REPLY transcript block model + themed chat renderer
   (escapement.tui.transcript, task 006). Covers: chronological SENT/REPLY
   classification, system collapsible?/chars, streaming? on the pending reply,
   reply meta carry-through; and the renderer's width-correctness (no \\n / no
   bleed on multiline bodies), hairline-before-reply, collapsed-system preview,
   the expand flag, and the streaming cursor."
  (:require
    [clojure.java.io :as io]
    [clojure.string :as str]
    [escapement.tui.compositor :as cmp]
    [escapement.tui.theme :as theme]
    [escapement.tui.transcript :as transcript]
    [fulcro-spec.core :refer [=> assertions specification component]]))

(def ^:private ESC (str (char 27)))                          ;; SGR escape lead

(defn- has-esc? [s] (str/includes? (str s) ESC))

(defn- mk-handle
  "Minimal TUI handle: a synthetic single-invocation scrollback (request/system,
   one user turn, one assistant response) plus optional in-flight live text."
  [invokeid & {:keys [live-text live-tokens]}]
  {:state (atom
            (cond->
              {:scrollback
               [{:source invokeid
                 :ev {:event :llm/request :ts 1000
                      :data {:system-preview "You are a helpful assistant. Be terse."
                             :model "gemma3:1b"}}}
                {:source invokeid
                 :ev {:event :llm/user-message :ts 2000
                      :data {:text "Write a haiku about snow."}}}
                {:source invokeid
                 :ev {:event :llm/tool-result :ts 2500
                      :data {:tool "get_haiku" :content-preview "{\"haiku\":[\"a\"]}"}}}
                {:source invokeid
                 :ev {:event :llm/response :ts 3000
                      :data {:stop-reason :end_turn
                             :usage {:input-tokens 12 :output-tokens 17}
                             :output-tps 42.0 :model "gemma3:1b"
                             :content [{:type :text :text "First snow falling soft on the silent winter field a hush over all"}]}}}]
               :term-w 80}
              live-text
              (assoc-in [:live invokeid :sessions]
                {:s1 {:status :streaming :text live-text
                      :tokens (or live-tokens 5) :model "gemma3:1b"}})))
   :env (atom {:escapement/session-dir "/tmp/does-not-exist-xyz"})})

(specification "transcript-blocks — pure SENT/REPLY model"
  (let [h      (mk-handle "poet-1")
        blocks (transcript/transcript-blocks h "poet-1")]
    (component "classification + order"
      (assertions
        "system/user/tool are SENT, assistant is REPLY, in chronological order"
        (mapv (juxt :dir :label) blocks)
        => [[:sent "system"] [:sent "user"] [:sent "tool"] [:reply "assistant"]]))
    (component "system block"
      (let [sys (first blocks)]
        (assertions
          "is collapsible with a char count"
          (:collapsible? sys) => true
          (number? (:chars (:meta sys))) => true
          (pos? (:chars (:meta sys))) => true)))
    (component "reply meta"
      (let [rep (last blocks)]
        (assertions
          "carries stop reason + token in/out + t/s"
          (get-in rep [:meta :stop]) => :end_turn
          (get-in rep [:meta :in])   => 12
          (get-in rep [:meta :out])  => 17
          (get-in rep [:meta :tps])  => 42.0
          "not streaming for a finalized response"
          (boolean (get-in rep [:meta :streaming?])) => false))))
  (component "in-flight reply"
    (let [h      (mk-handle "poet-2" :live-text "Line 1: the opening" :live-tokens 14)
          blocks (transcript/transcript-blocks h "poet-2")
          stream (last blocks)]
      (assertions
        "the pending reply is a streaming? REPLY block with the growing body"
        (:dir stream) => :reply
        (get-in stream [:meta :streaming?]) => true
        (str/includes? (:body stream) "the opening") => true)))
  (component "defensive"
    (assertions
      "missing/blank handle never throws; yields a vector"
      (vector? (transcript/transcript-blocks {:state (atom {})} "x")) => true)))

(defn- no-bleed?
  "Every line has no newline and display-width exactly = iw."
  [lines iw]
  (every? (fn [ln]
            (and (not (str/includes? ln "\n"))
              (= iw (cmp/display-width ln))))
    lines))

(specification "transcript-lines — themed renderer (no-bleed, hairline, collapse)"
  (let [h      (mk-handle "poet-1")
        blocks (transcript/transcript-blocks h "poet-1")
        none   (theme/theme-for :none)
        iw     50
        lines  (transcript/transcript-lines blocks none iw {:expanded? false})]
    (component "width correctness (no bleed, no embedded newlines)"
      (assertions
        "every emitted line is exactly iw columns and newline-free"
        (no-bleed? lines iw) => true
        "long body wraps across multiple lines"
        (> (count lines) (count blocks)) => true))
    (component "structure"
      (let [text (str/join "\n" (mapv str/trimr lines))]
        (assertions
          "SENT and REPLY headers are present"
          (str/includes? text "SENT") => true
          (str/includes? text "REPLY") => true
          "a hairline rule of ─ precedes the reply"
          (str/includes? text "─────") => true)))
    (component "system block renders its FULL body (never a preview)"
      (let [text (str/join "\n" lines)]
        (assertions
          "system header still advertises a char count (info only)"
          (str/includes? text "chars") => true)))
    (component "NO_COLOR ⇒ zero escapes"
      (assertions
        "no SGR escape under the :none theme"
        (boolean (some has-esc? lines)) => false)))
  (component "expand flag shows full system body"
    (let [h      (mk-handle "poet-1")
          blocks (transcript/transcript-blocks h "poet-1")
          none   (theme/theme-for :none)
          collapsed (transcript/transcript-lines blocks none 50 {:expanded? false})
          expanded  (transcript/transcript-lines blocks none 50 {:expanded? true})]
      (assertions
        "expanding the system block yields at least as many lines"
        (>= (count expanded) (count collapsed)) => true)))
  (component "streaming cursor on the in-flight body"
    (let [h      (mk-handle "poet-2" :live-text "tokens arriving now into the buffer")
          blocks (transcript/transcript-blocks h "poet-2")
          none   (theme/theme-for :none)
          lines  (transcript/transcript-lines blocks none 60 {:expanded? false})
          text   (str/join "\n" lines)]
      (assertions
        "a ▏ cursor appears in the rendered stream"
        (str/includes? text "▏") => true
        "still width-correct with the cursor"
        (no-bleed? lines 60) => true))))

(specification "themed (colored) renderer emits escapes"
  (let [h      (mk-handle "poet-1")
        blocks (transcript/transcript-blocks h "poet-1")
        th     (theme/theme-for :256)
        lines  (transcript/transcript-lines blocks th 60 {:expanded? false})]
    (assertions
      "under a color theme, headers carry SGR escapes"
      (boolean (some has-esc? lines)) => true
      "and remain width-correct (escapes are zero-width)"
      (no-bleed? lines 60) => true)))

(defn- mk-long-system-handle
  "Handle whose system prompt is long + multi-line, so we can prove the FULL
   body renders (wrapped), not a 1-line preview."
  [invokeid system]
  {:state (atom
            {:scrollback
             [{:source invokeid
               :ev {:event :llm/request :ts 1000
                    :data {:system-preview system :model "gemma3:1b"}}}
              {:source invokeid
               :ev {:event :llm/response :ts 3000
                    :data {:stop-reason :end_turn
                           :usage {:input-tokens 9 :output-tokens 3}
                           :content [{:type :text :text "ok"}]}}}]
             :term-w 80})
   :env (atom {:escapement/session-dir "/tmp/does-not-exist-xyz"})})

(specification "ISSUE 1 — full system body, never truncated to a preview"
  (let [;; long multi-line system prompt; every distinct sentinel must appear
        system  (str "SYSLINE-ALPHA the first directive sentence here. "
                  "SYSLINE-BETA a second very long instruction that should "
                  "wrap across several physical lines because it exceeds the "
                  "interior width by a comfortable margin and keeps going. "
                  "SYSLINE-GAMMA a third paragraph.\n"
                  "SYSLINE-DELTA on its own logical line.\n"
                  "SYSLINE-OMEGA the very last line of the system prompt.")
        h       (mk-long-system-handle "poet-9" system)
        blocks  (transcript/transcript-blocks h "poet-9")
        none    (theme/theme-for :none)
        iw      50
        lines   (transcript/transcript-lines blocks none iw {:expanded? false})
        text    (str/join "\n" lines)]
    (assertions
      "the FULL system body is present (first AND last sentinel), not a preview"
      (str/includes? text "SYSLINE-ALPHA") => true
      (str/includes? text "SYSLINE-DELTA") => true
      (str/includes? text "SYSLINE-OMEGA") => true
      "long body wraps across many lines (full content, wrapped)"
      (> (count lines) 8) => true
      "still width-correct (no bleed) and escape-free under :none"
      (no-bleed? lines iw) => true
      (boolean (some has-esc? lines)) => false)))

(specification "ISSUE 2 — SENT/REPLY direction tags stand out"
  (let [h       (mk-handle "poet-1")
        blocks  (transcript/transcript-blocks h "poet-1")
        none    (theme/theme-for :none)
        th      (theme/theme-for :256)
        ;; isolate the header lines (those carrying SENT / REPLY)
        sent-256  (first (filter #(str/includes? % "SENT")
                           (transcript/transcript-lines blocks th 70 {})))
        reply-256 (first (filter #(str/includes? % "REPLY")
                           (transcript/transcript-lines blocks th 70 {})))
        sent-none  (first (filter #(str/includes? % "SENT")
                            (transcript/transcript-lines blocks none 70 {})))
        reply-none (first (filter #(str/includes? % "REPLY")
                            (transcript/transcript-lines blocks none 70 {})))]
    (assertions
      "under :256 the SENT label carries SGR styling (bold/background)"
      (has-esc? sent-256) => true
      (has-esc? reply-256) => true
      "SENT and REPLY use DISTINCT tag styling (different bg codes)"
      (= (:sent-tag th) (:reply-tag th)) => false
      "under :none zero escapes, but the SENT/REPLY text remains"
      (has-esc? sent-none) => false
      (has-esc? reply-none) => false
      (str/includes? sent-none "SENT") => true
      (str/includes? reply-none "REPLY") => true)))

(specification "wrap-display — display-width word wrap"
  (assertions
    "wraps long text without exceeding width (visible cols)"
    (every? #(<= (cmp/display-width %) 20)
      (transcript/wrap-display 20 "  "
        "the quick brown fox jumps over the lazy dog repeatedly today")) => true
    "blank input yields no lines"
    (transcript/wrap-display 20 "  " "   ") => []
    "preserves the leading indent on each physical line"
    (every? #(str/starts-with? % "  ")
      (transcript/wrap-display 20 "  " "alpha beta gamma delta epsilon zeta eta")) => true))

(specification "back-compat fmt-transcript-event"
  (assertions
    "folds non-turn events to empty string"
    (transcript/fmt-transcript-event nil {:event :llm/start :ts 0 :data {}}) => ""
    "renders a user-message as a SENT turn"
    (str/includes?
      (transcript/fmt-transcript-event nil {:event :llm/user-message :ts 0
                                            :data {:text "hello"}})
      "hello") => true))

(specification "invocation-transcript-text — inspector pager seam (plain)"
  (let [h   (mk-handle "poet-1")
        txt (transcript/invocation-transcript-text h "poet-1")]
    (assertions
      "non-empty string"
      (and (string? txt) (pos? (count txt))) => true
      "contains both lanes and the body content"
      (str/includes? txt "SENT") => true
      (str/includes? txt "REPLY") => true
      (str/includes? txt "haiku") => true
      "is escape-free (survives the pager re-wrap)"
      (has-esc? txt) => false)))

(specification "blob reader cache — content-addressed blobs read once (BUG 2C)"
  (let [dir   (str (System/getProperty "java.io.tmpdir") "/esc-blob-cache-test")
        _     (.mkdirs (io/file dir))
        ref   "turns/0/response.edn"
        f     (io/file (str dir "/" ref))
        _     (.mkdirs (.getParentFile f))
        read! @#'transcript/read-blob-text]
    (reset! @#'transcript/blob-cache {})
    (spit f (pr-str "FIRST"))
    (assertions
      "first read returns the on-disk content"
      (read! dir ref) => "FIRST")
    ;; Mutate the file underneath — a memoized reader must NOT see the change
    ;; (blobs are append-only/content-addressed, so caching the first parse is
    ;; correct and the per-frame disk re-read is eliminated).
    (spit f (pr-str "SECOND"))
    (assertions
      "second read is served from cache (no disk re-read)"
      (read! dir ref) => "FIRST"
      "a missing blob caches nil rather than re-attempting"
      (read! dir "turns/9/missing.edn") => nil)))

File: tui/opentui/test/unit/transcript.test.ts
import { describe, expect, test } from "bun:test";
import { transcriptBlocks, type BlobReaders } from "../../src/domain/transcript";
import { liveAgg } from "../../src/domain/aggregate";
import { reduceFrames, initialDomainState } from "../../src/domain/store";
import { decodeFrame } from "../../src/transport/wire";
import type {
  EventLike,
  LiveAgg,
  ScrollbackEntry,
} from "../../src/domain/types";
import type { ForwardFrame } from "../../src/transport/wire";

const IID = "planner";
const TS = 1780798849000; // fixed -> deterministic hms

/** A scrollback entry wrapping an event keyed to a source. */
function entry(source: string, ev: EventLike): ScrollbackEntry {
  return { source, glyph: "·", summary: "", ev };
}
function ev(event: string, data: Record<string, unknown>, ts = TS): EventLike {
  return { event, ts, data };
}

/** Empty live rollup (no in-flight reply). */
const NO_LIVE: LiveAgg = {
  tokens: 0,
  tps: 0,
  n: 0,
  "n-active": 0,
  "n-done": 0,
  "last-ts": 0,
};

describe("transcriptBlocks — SENT/REPLY block model", () => {
  test("system request -> a collapsible SENT system block (from system-preview)", () => {
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry(IID, ev("llm/request", { invokeid: IID, "system-preview": "You are a poet." })),
      ],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(blocks).toHaveLength(1);
    expect(blocks[0]).toMatchObject({
      dir: "sent",
      label: "system",
      role: IID,
      body: "You are a poet.",
      "collapsible?": true,
    });
    expect(blocks[0]!.meta.chars).toBe("You are a poet.".length);
  });

  test("user-message -> SENT user block", () => {
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [entry(IID, ev("llm/user-message", { invokeid: IID, text: "Write a haiku" }))],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(blocks).toHaveLength(1);
    expect(blocks[0]).toMatchObject({ dir: "sent", label: "user", body: "Write a haiku" });
  });

  test("tool-result -> SENT tool block; is-error appends (ERROR)", () => {
    const ok = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry(IID, ev("llm/tool-result", { invokeid: IID, tool: "read_file", "content-preview": "contents" })),
      ],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(ok[0]).toMatchObject({ dir: "sent", label: "tool", sublabel: "read_file", body: "contents" });

    const err = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry(IID, ev("llm/tool-result", { invokeid: IID, tool: "read_file", "is-error": true, "content-preview": "boom" })),
      ],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(err[0]!.sublabel).toBe("read_file (ERROR)");
  });

  test("response with text + thinking + tool_use -> ordered REPLY blocks; meta on first", () => {
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry(
          IID,
          ev("llm/response", {
            invokeid: IID,
            "stop-reason": "end_turn",
            "output-tps": 68.08,
            usage: { "input-tokens": 12, "output-tokens": 30 },
            content: [
              { type: "thinking", thinking: "let me think" },
              { type: "text", text: "Cherry blossoms fall" },
              { type: "tool_use", name: "search", input: { q: "haiku" } },
            ],
          }),
        ),
      ],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(blocks.map((b) => [b.dir, b.label, b.sublabel])).toEqual([
      ["reply", "assistant", "thinking"],
      ["reply", "assistant", null],
      ["reply", "assistant", `search(${'{ "q": "haiku" }'})`], // collapseWs of pretty JSON
    ]);
    // meta attaches to the FIRST reply block only.
    expect(blocks[0]!.meta).toMatchObject({ stop: "end_turn", in: 12, out: 30, tps: 68.08 });
    expect(blocks[1]!.meta).toEqual({});
  });

  test("empty thinking block is dropped; meta promotes to the text block", () => {
    // Encrypted-reasoning providers (e.g. OpenAI) return a signature-only
    // thinking block with empty plaintext — nothing to render.
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry(
          IID,
          ev("llm/response", {
            invokeid: IID,
            "stop-reason": "end_turn",
            "output-tps": 54.3,
            usage: { "input-tokens": 2275, "output-tokens": 2156 },
            content: [
              { type: "thinking", thinking: "" },
              { type: "text", text: "# tournament-summary.md" },
            ],
          }),
        ),
      ],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(blocks.map((b) => [b.dir, b.label, b.sublabel])).toEqual([
      ["reply", "assistant", null],
    ]);
    // usage meta lands on the surviving text block.
    expect(blocks[0]!.meta).toMatchObject({ stop: "end_turn", in: 2275, out: 2156, tps: 54.3 });
  });

  test("tool_use sublabel is collapsed to one <=60-char accent line", () => {
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry(
          IID,
          ev("llm/response", {
            invokeid: IID,
            content: [{ type: "tool_use", name: "f", input: { big: "x".repeat(200) } }],
          }),
        ),
      ],
      live: NO_LIVE,
      now: () => TS,
    });
    const sub = blocks[0]!.sublabel!;
    // "f(" + collapsed args (<=60, trailing …) + ")"
    expect(sub.startsWith("f(")).toBe(true);
    expect(sub.endsWith(")")).toBe(true);
    expect(sub).toContain("…");
  });

  test("response with empty content -> one bare reply header carrying meta", () => {
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry(IID, ev("llm/response", { invokeid: IID, "stop-reason": "end_turn", content: [] })),
      ],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(blocks).toHaveLength(1);
    expect(blocks[0]).toMatchObject({ dir: "reply", label: "assistant", sublabel: null, body: "" });
    expect(blocks[0]!.meta.stop).toBe("end_turn");
  });

  test("llm/error -> a REPLY error block carrying the message", () => {
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [entry(IID, ev("llm/error", { invokeid: IID, message: "model crashed" }))],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(blocks[0]).toMatchObject({ dir: "reply", label: "error", body: "model crashed" });
  });

  test("in-flight streaming reply appended when live.text is non-blank", () => {
    const live: LiveAgg = { ...NO_LIVE, tokens: 7, text: "partial token stream" };
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [entry(IID, ev("llm/user-message", { invokeid: IID, text: "go" }))],
      live,
      now: () => TS,
    });
    const last = blocks[blocks.length - 1]!;
    expect(last).toMatchObject({ dir: "reply", label: "assistant", body: "partial token stream" });
    expect(last.meta["streaming?"]).toBe(true);
    expect(last.meta.out).toBe(7);
    expect(last.meta.tps).toBeNull(); // live-agg has no output-tps
  });

  test("blank/whitespace live.text adds NO in-flight block", () => {
    const live: LiveAgg = { ...NO_LIVE, text: "   " };
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [entry(IID, ev("llm/user-message", { invokeid: IID, text: "go" }))],
      live,
      now: () => TS,
    });
    expect(blocks.every((b) => b.meta["streaming?"] !== true)).toBe(true);
  });

  test("only this invokeid's scrollback contributes (other sources filtered)", () => {
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry("other", ev("llm/user-message", { invokeid: "other", text: "not mine" })),
        entry(IID, ev("llm/user-message", { invokeid: IID, text: "mine" })),
      ],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(blocks).toHaveLength(1);
    expect(blocks[0]!.body).toBe("mine");
  });

  test("one event shared across several scrollback entries is rendered ONCE (dedup by ev)", () => {
    const shared = ev("llm/user-message", { invokeid: IID, text: "once" });
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [entry(IID, shared), entry(IID, shared), entry(IID, shared)],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(blocks).toHaveLength(1);
  });

  test("full SENT->REPLY structure is chronological with system first", () => {
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry(IID, ev("llm/request", { invokeid: IID, "system-preview": "sys" }, TS)),
        entry(IID, ev("llm/user-message", { invokeid: IID, text: "u" }, TS + 1)),
        entry(IID, ev("llm/response", { invokeid: IID, content: [{ type: "text", text: "r" }] }, TS + 2)),
      ],
      live: NO_LIVE,
      now: () => TS,
    });
    expect(blocks.map((b) => [b.dir, b.label])).toEqual([
      ["sent", "system"],
      ["sent", "user"],
      ["reply", "assistant"],
    ]);
  });

  test("blob readers (when injected) override inline previews", () => {
    const blobs: BlobReaders = {
      system: () => "FULL SYSTEM PROMPT FROM BLOB",
      content: () => [{ type: "text", text: "FULL ASSISTANT FROM BLOB" }],
    };
    const blocks = transcriptBlocks({
      invokeid: IID,
      scrollback: [
        entry(IID, ev("llm/request", { invokeid: IID, "io/ref": "r1", "system-preview": "snippet" })),
        entry(IID, ev("llm/response", { invokeid: IID, "io/ref": "r2", content: [{ type: "text", text: "snippet" }] })),
      ],
      live: NO_LIVE,
      blobs,
      now: () => TS,
    });
    expect(blocks[0]!.body).toBe("FULL SYSTEM PROMPT FROM BLOB");
    expect(blocks[1]!.body).toBe("FULL ASSISTANT FROM BLOB");
  });
});

describe("transcriptBlocks — over the haiku fixture", () => {
  const text = require("fs").readFileSync(
    require("path").join(import.meta.dir, "../fixtures/haiku-sample.jsonl"),
    "utf8",
  ) as string;
  const frames = text
    .split("\n")
    .map((l: string) => l.trim())
    .filter((l: string) => l.length > 0)
    .map((l: string) => decodeFrame(l))
    .filter((f): f is ForwardFrame => f != null);
  const state = reduceFrames(frames, initialDomainState());

  test("planner has a coherent SENT->REPLY transcript with a response", () => {
    const blocks = transcriptBlocks({
      invokeid: "planner",
      scrollback: state.scrollback,
      live: liveAgg(state.live["planner"]?.sessions),
      now: () => TS,
    });
    expect(blocks.length).toBeGreaterThan(0);
    expect(blocks.some((b) => b.dir === "reply")).toBe(true);
  });

  test("poet2 error went to the :error lane, not the invokeid source -> 0 blocks", () => {
    const blocks = transcriptBlocks({
      invokeid: "poet2",
      scrollback: state.scrollback,
      live: liveAgg(state.live["poet2"]?.sessions),
      now: () => TS,
    });
    expect(blocks).toHaveLength(0);
  });
});
